### PR TITLE
add kube-dns for docker k8s

### PR DESCRIPTION
**Pull request by WIZARD-CXY**
@brendandburns Just look at the content and layout. If they're ok, I will fix the file url.

### CODE REVIEW

**Comment by WIZARD-CXY** [comment]
plus make it more readable.

**Comment by WIZARD-CXY** [comment]
@brendandburns Today I want to finish this dns part by using v1 api only to find current 0.18.2 hyperkube doesn't fully support it yet.
When I run `kubectl --namespace=kube-system create -f skydns-rc.yaml`  kubectl complains like below
`
Error from server: the server could not find the requested resource
`
While I use v1beta3 api, it's ok. So I think maybe it's because server side is old.Will you update to 0.19.3 or higher ?

**Comment by brendandburns** [comment]
@WIZARD-CXY it is updated to 0.21.2, I will update to 1.0.1 shortly.

Thanks!
--brendan

**Comment by WIZARD-CXY** [comment]
Roger Roger

**Comment by WIZARD-CXY** [comment]
@brendandburns  One thing I am wondering is whether we use https to get to apiserver or just http 8080 port for kube-dns. If we use https, we need to generate some certs and keys for apiserver and controller-manager

**Comment by brendandburns** [comment]
@WIZARD-CXY I think HTTP is ok for this example.  I'd like to get it to HTTPS eventually, but I'd rather err on the side of simplicity.

**Comment by WIZARD-CXY** [comment]
I'm not sure the v8 dns template can work with http. Let me have a try.even if it can't, I can set a flag to let kube2sky connect to apiserver 8080 http port.

May the force be with you

**Comment by WIZARD-CXY** [comment]
@brendandburns I just deployed a 0.21.2 docker-based k8s . And use a v8 dns template https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/addons/dns/skydns-rc.yaml.in

It isn't working. The kube2sky complains as follows

```
E0726 12:31:17.087853       1 reflector.go:136] Failed to list *api.Service: couldn't get version/kind; json parse error: invalid character '<' looking for beginning of value
E0726 12:31:17.088168       1 reflector.go:136] Failed to list *api.Endpoints: couldn't get version/kind; json parse error: invalid character '<' looking for beginning of value
```

In addition to that, from its log, 

```
I0726 12:31:16.076993       1 kube2sky.go:389] Etcd server found: http://127.0.0.1:4001
I0726 12:31:17.083316       1 kube2sky.go:455] Using http://localhost:8080 for kubernetes master
I0726 12:31:17.083361       1 kube2sky.go:456] Using kubernetes API
```
So it definitely can't reach apiserver since it uses localhost in his net namespace...

**Comment by dalanlan** [comment]
@googlebot  confirm

File: docs/getting-started-guides/docker-multinode/worker.md [hunk 1]
```diff
@@ -146,7 +146,7 @@ systemctl start docker
 Again this is similar to the above, but the `--api-servers` now points to the master we set up in the beginning.
 
 ```sh
-sudo docker run --net=host -d -v /var/run/docker.sock:/var/run/docker.sock  gcr.io/google_containers/hyperkube:v0.21.2 /hyperkube kubelet --api-servers=http://${MASTER_IP}:8080 --v=2 --address=0.0.0.0 --enable-server --hostname-override=$(hostname -i)
+sudo docker run --net=host -d -v /var/run/docker.sock:/var/run/docker.sock  gcr.io/google_containers/hyperkube::v1.0.1 /hyperkube kubelet --api-servers=http://${MASTER_IP}:8080 --v=2 --address=0.0.0.0 --enable-server --hostname-override=$(hostname -i --cluster-dns=10.0.0.10 --cluster-domain=cluster.local)
```
Comment (dalanlan) [inline docs/getting-started-guides/docker-multinode/worker.md]: Fatal errors.

Comment (WIZARD-CXY) [inline docs/getting-started-guides/docker-multinode/worker.md]: Aoh thanks.

**Comment by WIZARD-CXY** [comment]
@dalanlan would you kindly review it?

**Comment by WIZARD-CXY** [comment]
@brendandburns Sorry for the late update. Finally got time to finish this work.
Since we now use http api-server, the dns setting is a little tricky. I manually add the `kube_master_url` option for kube2sky to let it find the api-server. Otherwise it won't work. 
@thockin for good advice, if I miss something important.

Tested OK!

**Comment by WIZARD-CXY** [comment]
@brendandburns will you review it when you are not busy?

File: docs/getting-started-guides/docker-multinode/master.md [hunk 2]
```diff
@@ -196,7 +198,7 @@ If all else fails, ask questions on IRC at [#google-containers](http://webchat.f
 
 ### Next steps
 
-Move on to [adding one or more workers](worker.md)
+Move on to [adding one or more workers](worker.md) or [deply a dns](deployDNS.md)
```
Comment (brendandburns) [inline docs/getting-started-guides/docker-multinode/master.md]: s/deply/deploy/

File: docs/getting-started-guides/docker-multinode/worker.md [hunk 3]
```diff
@@ -146,7 +146,7 @@ systemctl start docker
 Again this is similar to the above, but the `--api-servers` now points to the master we set up in the beginning.
 
 ```sh
-sudo docker run --net=host -d -v /var/run/docker.sock:/var/run/docker.sock  gcr.io/google_containers/hyperkube:v0.21.2 /hyperkube kubelet --api-servers=http://${MASTER_IP}:8080 --v=2 --address=0.0.0.0 --enable-server --hostname-override=$(hostname -i)
+sudo docker run --net=host -d -v /var/run/docker.sock:/var/run/docker.sock  gcr.io/google_containers/hyperkube::v1.0.1 /hyperkube kubelet --api-servers=http://${MASTER_IP}:8080 --v=2 --address=0.0.0.0 --enable-server --hostname-override=$(hostname -i) --cluster-dns=10.0.0.10 --cluster-domain=cluster.local
```
Comment (brendandburns) [inline docs/getting-started-guides/docker-multinode/worker.md]: you have an extra ':' in `...hyperkube::v1.0.1`

File: docs/getting-started-guides/docker-multinode/deployDNS.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+### Use `kubectl` to create skydns rc and service
+
+
+```
+$ kubectl -s "$KUBE_SERVER:8080" --namespace=kube-system create -f ./skydns-rc.yaml
+
+$ kubectl -s "$KUBE_SERVER:8080" --namespace=kube-system create -f ./skydns-svc.yaml
+```
+
+### Test if dns works
```
Comment (brendandburns) [inline docs/getting-started-guides/docker-multinode/deployDNS.md]: s/dns/DNS/

**Comment by brendandburns** [comment]
@WIZARD-CXY Small comments, thanks for getting this working!

**Comment by WIZARD-CXY** [comment]
@brendandburns updated according to your comments, please review it.

**Comment by WIZARD-CXY** [comment]
@brendandburns Please give it a review, I think some guys are waiting for this update

**Comment by WIZARD-CXY** [comment]
@brendandburns will you please give it a review?

**Comment by brendandburns** [comment]
sorry for the delay.  LGTM

Thanks!

**Comment by WIZARD-CXY** [comment]
👌🏻

File: docs/getting-started-guides/docker-multinode/deployDNS.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+$ export DNS_SERVER_IP=10.0.0.10  # specify in startup parameter `--cluster-dns` for containerized kubelet 
+
+$ export KUBE_SERVER=10.10.103.250 # your master server ip, you may change it
+```
+
+### Replace the correponding value in the template.
+
+```
+$ sed -e "s/{{ pillar\['dns_replicas'\] }}/${DNS_REPLICAS}/g;s/{{ pillar\['dns_domain'\] }}/${DNS_DOMAIN}/g;s/{kube_server_url}/${KUBE_SERVER}/g;" skydns-rc.yaml.in > ./skydns-rc.yaml
```
Comment (lvlv) [inline docs/getting-started-guides/docker-multinode/deployDNS.md]: Just wondering, why we need KUBE_SERVER? skydns should be able to access master through https://${KUBERNETES_SERVICE_HOST}

Comment (WIZARD-CXY) [inline docs/getting-started-guides/docker-multinode/deployDNS.md]: you may read the previous conversation

May the force be with you

**Pull request merged**